system-addon/lib/LinksCache.jsm

```diff
@@ -78,26 +80,31 @@ this.LinksCache = class LinksCache {
           }
         }
 
-        // Make a shallow copy of each resulting link to allow direct edits
+        // Copy each link to avoid modifying original source objects below
         const copied = (await this.linkGetter(options)).map(link => link &&
           Object.assign({}, link));
 
-        // Migrate data to the new link if we have an old link
+        // Modify each of the link objects by migrating or adding data
         for (const newLink of copied) {
-          if (newLink) {
-            const oldLink = toMigrate.get(newLink.url);
-            if (oldLink) {
-              this.migrator(oldLink, newLink);
-            }
+          if (!newLink) {
```

Comment:
Is there a reason we can't filter these out before the map on line 84?

Reply:
We need to keep holes for pinned links

Reply:
I suppose I could clean this up a little to avoid the `map` then `for of` both checking for object existing.

Reply:
Saved a few lines with better clarity (added comment about holes) `1 file changed, 9 insertions(+), 13 deletions(-)`